File: 2. Week 2 - Scatter Plot/data/extract_data.R
# Week 2 - Data Extraction (R version)
# Goal: Calculate Distance and Pace for Scatter Plot

library(tidyverse)

input_path <- "../../data/runs_only_redacted.csv"
output_path <- "distance_vs_pace.csv"

if (file.exists(input_path)) {
  df <- read_csv(input_path, show_col_types = FALSE)
  
  # Python script logic:
  # Distance_km = Distance
  # AveragePace_min_km = Time (min) / Distance (km)
  # But we need time in minutes. 
  # If 'Time' is in seconds (common in GPS watches), divide by 60.
  # Let's check headers. Usually 'Time' or 'Duration'. 
  # Assuming 'Time' exists and is in minutes based on Python script viewing.
  # "df['Time'] / 60" in python implies 'Time' column was seconds? 
  # Wait, let's verify column names from previous Context or assume standard:
  # "Time" (seconds) -> / 60 -> minutes.
  
  processed_df <- df %>%
    mutate(
      Distance_km = round(Distance, 2),
      
      # Python script: (df['Time'] / 60) / df['Distance']
      # So 'Time' is in seconds.
      AveragePace_min_km = round((Time / 60) / Distance, 2)
    ) %>%
    select(Distance_km, AveragePace_min_km) %>%
    filter(!is.na(Distance_km) & !is.na(AveragePace_min_km)) %>%
    filter(is.finite(AveragePace_min_km)) # Clean infinite pace (div by zero)
  
  write_csv(processed_df, output_path)
  print(paste("Saved data to", output_path))
  
} else {
  stop("Input file not found")
}

File: 2. Week 2 - Scatter Plot/data/extract_data.py
import pandas as pd
import os

def extract_data():
    input_path = '../../data/runs_only_redacted.csv'
    output_path = 'distance_vs_pace.csv'
    
    if not os.path.exists(input_path):
        print(f"Error: Input file not found at {input_path}")
        return

    try:
        df = pd.read_csv(input_path)
        
        # Check columns
        required_cols = ['distance', 'average_speed'] # inferred from previous checks
        # If 'average_speed' (m/s) is present, we calculate pace (min/km).
        # Pace = 16.6667 / Speed
        
        # Normalize columns if needed
        df.columns = [c.lower() for c in df.columns]
        
        if 'distance' not in df.columns:
            print("Error: 'distance' column not found")
            return
            
        # Extract and Calculate
        output_df = pd.DataFrame()
        output_df['Distance'] = df['distance'] / 1000 # Convert m to km
        
        if 'average_pace' in df.columns:
             output_df['AveragePace'] = df['average_pace']
        elif 'average_speed' in df.columns:
             # Avoid division by zero
             output_df['AveragePace'] = df['average_speed'].apply(lambda x: 16.666666667 / x if x > 0 else 0)
        else:
            print("Error: check pace/speed columns")
            return
            
        # Round and Filter
        output_df = output_df.round(2)
        
        # Filter out invalid 0s
        output_df = output_df[(output_df['Distance'] > 0) & (output_df['AveragePace'] > 0)]
        
        output_df.to_csv(output_path, index=False)
        print(f"Successfully created {output_path}")
        print(output_df.head())

    except Exception as e:
        print(f"Error: {e}")

if __name__ == "__main__":
    extract_data()
